src/log-store/src/raft_engine/log_store.rs

```diff
@@ -116,22 +116,71 @@ impl RaftEngineLogStore {
         )
     }
 
-    /// Checks if entry does not override the min index of namespace.
-    fn check_entry(&self, e: &EntryImpl) -> Result<()> {
-        if cfg!(debug_assertions) {
+    /// Converts entries to `LogBatch` and checks if entry ids are valid.
+    /// Returns the `LogBatch` converted along with the last entry id
+    /// to append in each namespace(region).
+    fn entries_to_batch(
+        &self,
+        entries: Vec<EntryImpl>,
+    ) -> Result<(LogBatch, HashMap<NamespaceId, EntryId>)> {
+        // Records the last entry id for each region's entries.
+        let mut entry_ids: HashMap<NamespaceId, BTreeSet<EntryId>> =
+            HashMap::with_capacity(entries.len());
+
+        let mut batch = LogBatch::with_capacity(entries.len());
+
+        for e in entries {
             let ns_id = e.namespace_id;
-            if let Some(first_index) = self.engine.first_index(ns_id) {
+            if !entry_ids.entry(ns_id).or_default().insert(e.id) {
+                return DuplicateLogIndexSnafu {
+                    region_id: ns_id,
+                    index: e.id,
+                }
+                .fail();
+            }
+            batch
+                .add_entries::<MessageType>(ns_id, &[e])
+                .context(AddEntryLogBatchSnafu)?;
+        }
+
+        let mut last_entry_ids = HashMap::with_capacity(entry_ids.len());
+        for (region, ids) in entry_ids {
+            let first_in_batch = *ids.first().unwrap();
+            let last_in_batch = *ids.last().unwrap();
+            ensure!(
+                (last_in_batch - first_in_batch) == ids.len() as u64 - 1,
+                DiscontinuousLogIndexSnafu {
+                    region_id: region,
+                    last_index: first_in_batch,
+                    attempt_index: last_in_batch
+                }
+            );
+
+            if let Some(first_index) = self.engine.first_index(region) {
                 ensure!(
-                    e.id() >= first_index,
+                    first_in_batch >= first_index,
```

Comment:
Shall we allow `first_in_batch == first_index`?

Reply:
> Shall we allow `first_in_batch == first_index`?

Mito cannot tolerate this, but raft-engine allows override non-compacted existing entries.